### PR TITLE
Credit Reviewers in Release Notes

### LINKED ISSUE #2921: Reviewers do not receive credit in release notes
We should update our release notes script to also examine GitHub to find out who contributed to code review.

### LINKED ISSUE #3095: Modify release note script to also generate list of reviewers
No piece of code can be included an a release without review, and as such reviewers should also be mentioned in the release notes.

### PR DESCRIPTION
**Pull request by kne42**
## Description
Credits reviewers and authors of merged PRs on GitHub.

It is recommended that `GH_TOKEN` be set to avoid running into problems with [rate limiting](https://developer.github.com/v3/#rate-limiting). One can be acquired at https://github.com/settings/tokens. While PRs and forks will not have access to the encrypted `GH_TOKEN` environment variable, the main branch should properly receive it which is important when deploying the docs.

## References
- Closes #2921 
- Closes #3095 

GitHub API Documentation:
- https://developer.github.com/v3/search/#search-issues
- https://help.github.com/articles/understanding-the-search-syntax/#query-for-dates
- https://developer.github.com/v3/pulls/reviews/
- https://developer.github.com/v3/users/#get-a-single-user
- https://developer.github.com/v3/#rate-limiting
- https://developer.github.com/v3/#oauth2-token-sent-as-a-parameter

HTTP Request Documentation:
- https://docs.python.org/2/library/urllib.html
- https://docs.python.org/3/library/urllib.request.html
- https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode

Git Documentation:
- https://git-scm.com/docs/pretty-formats
## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by kne42** [comment]
While it has the expected output on my computer, it doesn’t seem to be working properly on Travis. I will add a verbose debug mode and clean up the code more later to pinpoint the underlying issues.

**Comment by kne42** [comment]
Most changes will be made on my own branch, [kne42:debug-credit-reviewers](/kne42/scikit-image/tree/debug-credit-reviewers), for access to a customized Travis build. I will additionally display all credits since the last release.

**Comment by stefanv** [comment]
@kne42 I'm excited to see this PR!  Do you have an example of output?  Is it even worth trying this without a token?  It may be easier to just print an instruction on how to set up the token, and proceed with the rest.

**Comment by kne42** [comment]
@stefanv Ah yes, I forgot that we will be releasing a new version soon! I can have this finished by Friday.

**Comment by kne42** [comment]
Here is an example output:
```
Release HEAD~50 was on 2017-10-05T11:01:32+02:00

A total of 155 changes have been committed.

It contained the following 99 merged pull requests:
- 3100 : Optimize _probabilistic_hough_line function
- 3096 : Add `threshold_local` to `filters` module namespace
- 3093 : Force Appveyor to fail on failed tests
- 3085 : v0.13.x: Backport NumPy 1.14 compatibility
- 3076 : Allow convex_hull_image on empty images
- 3074 : Sanitizing marching_cubes_lewiner spacing input argument
- 3072 : Update dependencies and deprecations to fix Travis builds
- 3063 : For sparse CG solver, provide atol=0 keyword for SciPy >= 1.1
- 3056 : speedup img_as_float by making division multiplication and avoiding unecessary allocation
- 3050 : Switch to basis numpy int dtypes in dtype_range
- 3042 : Correct reference for the coins sample image
- 3040 : Fix examples not displaying figures
- 3039 : Fix invert tests
- 3038 : Fix RuntimeError: generator raised StopIteration on Python 3.7
- 3037 : Add __pycache__ and .cache to .gitignore.
- 3035 : Replace pow function calls in Cython modules to fix performance issues on Windows
- 3033 : Fix ValueError: Buffer dtype mismatch, expected 'int64_t' but got 'int' on win_amd64
- 3030 : Type dependent inversion
- 3029 : `draw.random_shapes` API improvements
- 3023 : Minor style and documentation updates for #2859
- 3018 : Add YDbDr colorspace conversion routines
- 3013 : Add citation information to README
- 3008 : Ensure stdev is always nonnegative in _mean_std
- 3004 : MAINT: PIL removed saving RGBA images as jpeg files
- 2999 : Add release notes for 0.13.1
- 2996 : Return empty list on flat images with hough_ellipse #2820
- 2979 : Bugfix LineModelND.residuals does not use the optional parameter `params`
- 2977 : Handle matplotlib 2.2 pre-release deprecations
- 2975 : Use correct NumPy version comparison in pytest configuration
- 2971 : Mark data_range as being a float

Made by the following committers [alphabetical by last name]:
- Norman Barker
- Leonid Bloch
- Benedikt Boecking
- François Boulogne
- Larry Bradley
- Matthew Brett
- Alex Chum
- Yannick Copin
- Nethanel Elzas
- Kira Evans
- Emmanuelle Gouillart
- Ben Hadfield
- Mark Harfouche
- Scott Heatwole
- Gregory R. Lee
- Guillaume Lemaitre
- Kevin Mader
- Jarrod Millman
- Pradyumna Narayana
- Juan Nunez-Iglesias
- Egor Panfilov
- Oleksandr Pavlyk
- Alex Rothberg
- Max Schambach
- Johannes Schönberger
- Johannes Schönberger
- Nelle Varoquaux
- Viraj
- Stefan van der Walt
- Thomas Walter
- Scott Warchal
- blochl
- cgohlke
- eepaillard
- goldsborough
- mikigom
- mutterer
- nrweir

Reviewed by the following reviewers [alphabetical by last name]:
- François Boulogne
- Peter Goldsborough
- Juan Nunez-Iglesias
- Stefan van der Walt
- Josh Warner
- Eric Wieser
```

**Comment by kne42** [comment]
Note that Johannes appears twice in the output because there are two visually similar unicode variations of his last name: one contains `111+776` while the other contains `246`.

**Comment by kne42** [comment]
@stefanv @soupault Script is now ready for use/review! It is encountering problems in certain Travis builds but I don't think we need to merge this before v0.14.0 so I can fix that later.

**Comment by soupault** [comment]
I'm not 100% confident, but the script might include authors of the approved, but not merged PRs to the list of contributors.

File: doc/release/contribs.py [hunk 1]
```diff
@@ -1,50 +1,193 @@
 #!/usr/bin/env python
+from __future__ import print_function
```
Comment (soupault) [inline doc/release/contribs.py]: Not necessary since we do not support Python 2 anymore.

File: doc/release/contribs.py [hunk 2]
```diff
@@ -1,50 +1,193 @@
...
+import os
 import sys
 import string
 import shlex
+import json
+from datetime import datetime
+import math
+import time
+
+if sys.version_info[0] < 3:
```
Comment (soupault) [inline doc/release/contribs.py]: Same.

File: doc/release/contribs.py [hunk 3]
```diff
@@ -1,50 +1,193 @@
...
+GH_TOKEN = os.environ.get('GH_TOKEN')
 
 if len(sys.argv) != 2:
-    print("Usage: ./contributors.py tag-of-previous-release")
+    print("Usage: ./contribs.py [--latest|tag-of-previous-release]")
     sys.exit(-1)
 
 tag = sys.argv[1]
 
+if GH_TOKEN is None:
```
Comment (soupault) [inline doc/release/contribs.py]: I'd add this also to `RELEASE.txt`.

File: doc/release/contribs.py [hunk 4]
```diff
@@ -1,50 +1,193 @@
...
+    # See https://developer.github.com/v3/users/#get-a-single-user
+    url = 'https://api.github.com/users/' + login
+    user = request(url)
+    return user
+
+
+def get_merged_pulls(user, repo, date, page=1, results=0):
+    # See https://developer.github.com/v3/search/#search-issues
+    # See https://help.github.com/articles/understanding-the-search-syntax/#query-for-dates
+    query = 'repo:%s/%s merged:>%s sort:created-asc' % (user, repo, date)
```
Comment (soupault) [inline doc/release/contribs.py]: Maybe better to use `.format` or even formatted strings `f'....'`.

File: doc/release/contribs.py [hunk 5]
```diff
@@ -1,50 +1,193 @@
...
-tag_date = call("git log -n1 --format='%%ci' %s" % tag)[0]
-print("Release %s was on %s\n" % (tag, tag_date))
+    count = len(merges['items'])
 
-merges = call("git log --since='%s' --merges --format='>>>%%B' --reverse" % tag_date)
-merges = [m for m in merges if m.strip()]
-merges = '\n'.join(merges).split('>>>')
-merges = [m.split('\n')[:2] for m in merges]
-merges = [m for m in merges if len(m) == 2 and m[1].strip()]
+    if results < merges['total_count']:
```
Comment (soupault) [inline doc/release/contribs.py]: Shoud be `while`?

File: doc/release/contribs.py [hunk 6]
```diff
@@ -1,50 +1,193 @@
...
+    reviews = get_reviews(GH_USER, GH_REPO, id)
+    for review in reviews:
+        try:
+            reviewer = review['user']['login']
+            if author == reviewer:  # author reviewing own PR
+                continue
+            if reviewer not in users:
+                name = get_user(reviewer).get('name')
+                if name is None:
+                    name = handle
```
Comment (soupault) [inline doc/release/contribs.py]: What is `handle`?

**Comment by stefanv** [comment]
Are there any outstanding issues with this PR, or is it ready to go?

**Comment by jni** [comment]
@kne42 would you mind updating to address @soupault's Py2 comments?

**Comment by stefanv** [comment]
@kne42 One last question: @soupault mentioned that it may pick up approved & not merged PR reviewers.  Is this an issue we should look into?

**Comment by kne42** [comment]
@stefanv @soupault 
No; reviewers and authors were only pulled from the results of [this query](https://github.com/search?q=repo:scikit-image/scikit-image+merged:%3E2017-09-27T09:38:27%2B10:00+sort:created-asc) which, as you can see, contains only merged pull requests as the query specifically filters out any non-merged pull requests.

**Comment by stefanv** [comment]
Good catch; I think reviewing your own PR should not count.

**Comment by soupault** [comment]
I've refactored the script to use `argparse`, a dedicated main loop, `.format` instead of `%s`, and fixed invalid format specifiers in these lines:
```
tag_date = call("git log -n1 --format='%cI' {}".format(tag))[0]
num_commits = call("git rev-list {}..HEAD --count".format(tag))[0]
committers = call("git log --since='{}' --format=%aN".format(tag_date))
```

**Comment by soupault** [comment]
@scikit-image/core anyone for the second review?

**Comment by jni** [comment]
@kne42 and @soupault thank you! Brilliant stuff.

**Pull request merged**